### ISSUE TITLE
Offset returns incorrect value in some circumstances

### ISSUE THREAD

**Issue by 0xxon** [zeek/spicy#710]
Given the following unit:

```
module Test;

public type Testing = unit {
	la: Sub(4);
};

type Sub = unit(length: uint32) {
	%random-access;

	a: bytes &size=2;
	z: bytes &size=length-cast<uint32>(self.offset());

	on %done {
		print self;
	}
};
```

I would expect a to have a length of 2 bytes and z to have a length of two bytes (since it is read two bytes into the stream with a passed length of 4 bytes). However, self.offset() returns a value of zero when reading z, resulting in the following output:

```
$ echo "1234567890" | spicy-driver test.spicy
[$a=b"12", $z=b"3456"]
```

You can work around this by introducing a dummy, zero-length field inbetween the two fields; with this self.offset() returns 2 and only reads two bytes in the example:

```
module Test;

public type Testing = unit {
	la: Sub(4);
};

type Sub = unit(length: uint32) {
	%random-access;

	a: bytes &size=2;
	: bytes &size=0;
	z: bytes &size=length-cast<uint32>(self.offset());

	on %done {
		print self;
	}
};
```

->

```
$ echo "1234567890" | spicy-driver test.spicy
[$a=b"12", $z=b"34"]
```

**Comment by 0xxon** [comment]
To add to this, because Robin just asked me - the offset seems to be correct in hooks for the elements. E.g.:

```
module Test;

public type Testing = unit {
	la: Sub(4);
};

type Sub = unit(length: uint32) {
	%random-access;

	a: bytes &size=2;
	z: bytes &size=length-cast<uint32>(self.offset());

	on a {
		print "a", self.offset();
	}
	on z {
		print "z", self.offset();
	}
	on %done {
		print self;
	}
};
```
Correctly outputs the offset values in the hooks, even though it reads too much:

```
$ echo "1234567890" | spicy-driver test.spicy
a, 0
z, 2
[$a=b"12", $z=b"3456"]
```

**Comment by bbannier** [comment]
Closed by #716.
